### PR TITLE
fix statick check errors

### PR DESCRIPTION
**Pull request by krasi-georgiev**
fixes: https://github.com/prometheus/tsdb/issues/434

the static check also found some actual bugs.

will run some bench test to check if there is any differences

Signed-off-by: Krasi Georgiev <kgeorgie@redhat.com>

### CODE REVIEW

**Comment by krasi-georgiev** [comment]
tsdb bench tests

```
benchcmp -mag old.txt new.txt 
benchmark                                                           old ns/op     new ns/op     delta
BenchmarkMergedSeriesSet/series=20000,blocks=4-32                   24834504      23299186      -6.18%
BenchmarkMergedSeriesSet/series=20000,blocks=2-32                   10946125      10454205      -4.49%
BenchmarkMergedSeriesSet/series=10000,blocks=4-32                   11822581      11335265      -4.12%
BenchmarkMergedSeriesSet/series=10000,blocks=2-32                   3897518       3741115       -4.01%
BenchmarkHeadStripeSeriesCreate-32                                  5173          4981          -3.71%
BenchmarkMergedSeriesSet/series=20000,blocks=16-32                  100931355     98822885      -2.09%
BenchmarkMergedSeriesSet/series=100,blocks=16-32                    508465        498747        -1.91%
BenchmarkMergedSeriesSet/series=20000,blocks=8-32                   48588915      47724934      -1.78%
BenchmarkMergedSeriesSet/series=10000,blocks=8-32                   24251144      23879048      -1.53%
BenchmarkMergedSeriesSet/series=100,blocks=8-32                     224188        226884        +1.20%
BenchmarkMergedSeriesSet/series=1000,blocks=8-32                    2377099       2355152       -0.92%
BenchmarkMergedSeriesSet/series=1000,blocks=4-32                    998721        1004779       +0.61%
BenchmarkMergedSeriesSet/series=1000,blocks=2-32                    327024        328910        +0.58%
BenchmarkMergedSeriesSet/series=1000,blocks=16-32                   5221600       5194657       -0.52%
BenchmarkPersistedQueries/series=10,samplesPerSeries=10000-32       9781145       9736872       -0.45%
BenchmarkMergedSeriesSet/series=1000,blocks=32-32                   11309230      11258176      -0.45%
BenchmarkMergedSeriesSet/series=100,blocks=2-32                     30679         30814         +0.44%
BenchmarkMergedSeriesSet/series=100,blocks=1-32                     910           914           +0.44%
BenchmarkPersistedQueries/series=100,samplesPerSeries=100000-32     962824202     966406732     +0.37%
BenchmarkMergedSeriesSet/series=10000,blocks=32-32                  109758490     109391035     -0.33%
BenchmarkPersistedQueries/series=10,samplesPerSeries=1000-32        1024498       1027770       +0.32%
BenchmarkMergedSeriesSet/series=10000,blocks=16-32                  50288056      50158292      -0.26%
BenchmarkMergedSeriesSet/series=100,blocks=32-32                    1091321       1089122       -0.20%
BenchmarkMergedSeriesSet/series=20000,blocks=1-32                   95705         95879         +0.18%
BenchmarkMergedSeriesSet/series=20000,blocks=32-32                  226768194     227144561     +0.17%
BenchmarkPersistedQueries/series=100,samplesPerSeries=1000-32       9995215       9982456       -0.13%
BenchmarkPersistedQueries/series=100,samplesPerSeries=10000-32      97006233      96903401      -0.11%
BenchmarkPersistedQueries/series=10,samplesPerSeries=100000-32      96283961      96384511      +0.10%
BenchmarkHeadStripeSeriesCreateParallel-32                          3919          3921          +0.05%
BenchmarkMergedSeriesSet/series=1000,blocks=1-32                    5308          5310          +0.04%
BenchmarkMergedSeriesSet/series=100,blocks=4-32                     92778         92760         -0.02%
BenchmarkHeadPostingForMatchers-32                                  94013         94000         -0.01%
BenchmarkMergedSeriesSet/series=10000,blocks=1-32                   48150         48146         -0.01%

benchmark                                                           old allocs     new allocs     delta
BenchmarkPersistedQueries/series=10,samplesPerSeries=1000-32        430            430            +0.00%
BenchmarkPersistedQueries/series=10,samplesPerSeries=10000-32       2683           2683           +0.00%
BenchmarkPersistedQueries/series=10,samplesPerSeries=100000-32      25187          25187          +0.00%
BenchmarkPersistedQueries/series=100,samplesPerSeries=1000-32       3687           3687           +0.00%
BenchmarkPersistedQueries/series=100,samplesPerSeries=10000-32      26191          26191          +0.00%
BenchmarkPersistedQueries/series=100,samplesPerSeries=100000-32     251202         251202         +0.00%

benchmark                                                           old bytes     new bytes     delta
BenchmarkPersistedQueries/series=10,samplesPerSeries=100000-32      1689296       1689337       +0.00%
BenchmarkPersistedQueries/series=10,samplesPerSeries=1000-32        68198         68199         +0.00%
BenchmarkPersistedQueries/series=10,samplesPerSeries=10000-32       210449        210452        +0.00%
BenchmarkPersistedQueries/series=100,samplesPerSeries=100000-32     15692155      15692323      +0.00%
BenchmarkPersistedQueries/series=100,samplesPerSeries=1000-32       304796        304799        +0.00%
BenchmarkPersistedQueries/series=100,samplesPerSeries=10000-32      1643215       1643223       +0.00%
```

**Comment by krasi-georgiev** [comment]
promql bench showed some decreased performance when fixing SA6002 so added a rule to ignore this check and reverted the changes.

File: querier.go [hunk 1]
```diff
@@ -620,11 +620,9 @@ func (s *populatedChunkSeries) Next() bool {
 				// This means that the chunk has be garbage collected. Remove it from the list.
 				if s.err == ErrNotFound {
 					s.err = nil
-
 					// Delete in-place.
-					chks = append(chks[:j], chks[j+1:]...)
+					s.chks = append(chks[:j], chks[j+1:]...)
```
Comment (codesome) [inline querier.go]: Nice catch, it is easy to mistake that `s.chks` will be assigned to `chks` later.

File: Makefile [hunk 2]
```diff
@@ -18,7 +18,10 @@ TSDB_BENCHMARK_NUM_METRICS ?= 1000
 TSDB_BENCHMARK_DATASET ?= "$(TSDB_PROJECT_DIR)/testdata/20kseries.json"
 TSDB_BENCHMARK_OUTPUT_DIR ?= "$(TSDB_CLI_DIR)/benchout"
 
-STATICCHECK_IGNORE =
+STATICCHECK_IGNORE = \
+	github.com/prometheus/tsdb/head.go:SA6002 \
```
Comment (bwplotka) [inline Makefile]: Worth a comment why?

Comment (brian-brazil) [reply Makefile]: We don't elsewhere.

Comment (krasi-georgiev) [reply Makefile]: yeah I think a comment is useful as I wasted 1 hour  on this one.

Comment (krasi-georgiev) [reply Makefile]: added

**Comment by krasi-georgiev** [comment]
not sure what's up with this travis error. Something with the go modules. runs fine locally.

**Comment by krasi-georgiev** [comment]
@simonpasquier have you seen this go mod in the failing travis error before?

**Comment by simonpasquier** [comment]
I think this should fix the problem:

https://github.com/prometheus/tsdb/pull/490/files/f93318a95a842c60085eedfd783e96bb4c6c1fec..5977601d942085ccb3dfd128156382eaf9b23231

I'm not sure about what's happening with staticcheck but pulling all dependencies (including test deps) before running `staticcheck` helps...

**Comment by krasi-georgiev** [comment]
Thanks. that seems to have fixed it, 
can you push these changes into my branch `fix-check-license-staticcheck`

**Pull request merged**